### PR TITLE
Introduce OutlierMode

### PR DESCRIPTION
**Pull request by AndreyAkinshin**

### CODE REVIEW

**Comment by AndreyAkinshin** [comment]
@adamsitnik, please review, but don't merge (I'm going to update the documentation as soon as you approve the naming).

File: src/BenchmarkDotNet/Mathematics/OutlierMode.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace BenchmarkDotNet.Mathematics
+{
+    public enum OutlierMode
+    {
+        None, OnlyUpper, OnlyLower, All
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Mathematics/OutlierMode.cs]: I would add some xml docs here (some users might not understand what is Upper and Lower in this context)

File: src/BenchmarkDotNet/Engines/EngineResolver.cs [hunk 2]
```diff
@@ -24,16 +25,16 @@ private EngineResolver()
...
             Register(AccuracyMode.EvaluateOverheadCharacteristic, () => true);
-            Register(AccuracyMode.RemoveOutliersCharacteristic, job =>
+            Register(AccuracyMode.OutlierModeCharacteristic, job =>
             {
                 var strategy = job.ResolveValue(RunMode.RunStrategyCharacteristic, this);
                 switch (strategy)
                 {
                     case RunStrategy.Throughput:
-                        return true;
+                        return OutlierMode.OnlyUpper;
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Engines/EngineResolver.cs]: @AndyAyersMS this change implements your request to remove the outliers that are above the upper fence (the ones below lower fence are now going to remain)

@AndreyAkinshin 👍

File: src/BenchmarkDotNet/Jobs/JobExtensions.cs [hunk 3]
```diff
@@ -52,7 +52,7 @@ public static class JobExtensions
         public static Job WithMinIterationTime(this Job job, TimeInterval value) => job.WithCore(j => j.Accuracy.MinIterationTime = value);
         public static Job WithMinInvokeCount(this Job job, int value) => job.WithCore(j => j.Accuracy.MinInvokeCount = value);
         public static Job WithEvaluateOverhead(this Job job, bool value) => job.WithCore(j => j.Accuracy.EvaluateOverhead = value);
-        public static Job WithRemoveOutliers(this Job job, bool value) => job.WithCore(j => j.Accuracy.RemoveOutliers = value);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Jobs/JobExtensions.cs]: @AndreyAkinshin I think that we could keep `WithRemoveOutliers` for backward compatibility and implement it to keep the old behavior. Just to have fewer breaking changes.

```cs
[Obsolete("Please use the new WithOutlierMode instead")]
public static Job WithRemoveOutliers(this Job job, bool value) 
    => job.WithOutlierMode(value ? OutlierMode.All : OutlierMode.None);
```

Comment (AndreyAkinshin) [reply src/BenchmarkDotNet/Jobs/JobExtensions.cs]: @adamsitnik fixed.

File: src/BenchmarkDotNet/Mathematics/OutlierMode.cs (hunk 1 above)
Comment (AndreyAkinshin) [reply src/BenchmarkDotNet/Mathematics/OutlierMode.cs]: Done.

File: src/BenchmarkDotNet/Jobs/JobExtensions.cs [hunk 4]
```diff
@@ -52,7 +52,9 @@ public static class JobExtensions
...
+        public static Job WithOutlierMode(this Job job, OutlierMode value) => job.WithCore(j => j.Accuracy.OutlierMode = value);
+        [Obsolete("Please use the new WithOutlierMode instead")]
+        public static Job WithRemoveOutliers(this Job job, bool value) => job.WithCore(j => j.Accuracy.OutlierMode = OutlierMode.OnlyUpper);
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Jobs/JobExtensions.cs]: @AndreyAkinshin you forgot the case where `value == false` (users don't want to remove the outliers)

Comment (AndreyAkinshin) [reply src/BenchmarkDotNet/Jobs/JobExtensions.cs]: Fixed.

**Pull request merged**